### ISSUE TITLE
Markdown.parse doesn't parse raw HTML properly.

### ISSUE THREAD

**Issue by bicycle1885** [JuliaLang/julia#17837]
I noticed `Markdown.parse` doesn't parse HTML tags properly while writing docs with `Documenter.jl` (xref: https://github.com/JuliaDocs/Documenter.jl/issues/176).

Most Markdown parsers support this feature, so I think `Base.Markdown` should do as well.

For example, two consecutive hyphens are recognized as an em dash as follows:
```
julia> Markdown.parse("<!-- comment -->")
  <!– comment –>

```

CC: @MichaelHatherly 

---

```
julia> versioninfo()
Julia Version 0.5.0-rc1+0
Commit cede539* (2016-08-04 08:48 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i5-4288U CPU @ 2.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)

```

**Comment by StefanKarpinski** [comment]
I fear that we're going to end up having a full HTML parser in Base :|

**Comment by eschnett** [comment]
Nice! Which Javascript library are you going to use?

**Comment by MichaelHatherly** [comment]
> I fear that we're going to end up having a full HTML parser in Base :|

Yeah, that's not something I'd like to end up happening. Most markdown parsers seem to just use some regex monstrosities to catch raw HTML, which appears to work alright.

**Comment by KristofferC** [comment]
Reminds me of this: http://stackoverflow.com/a/1732454

**Comment by bicycle1885** [comment]
How about using [CommonMark](http://commonmark.org/)? It already has the [libcmark](https://github.com/jgm/cmark) library that supports HTML tags.

**Comment by MichaelHatherly** [comment]
Does CommonMark support some form of table syntax yet @bicycle1885? From the last time I looked through the spec I didn't come across anything.

I think it would probably be a good idea to wrap libcmark anyway (at some point) even if it's just to make it easier to check how much of CommonMark we actually adhere to, which is most likely not much at the moment.

**Comment by bicycle1885** [comment]
No, CommonMark seems to be very conservative to add extensions like table syntax. I'm not sure but I think we can do some preprocessing to convert table syntax extension to HTML tables before passing a string to libcmark.

**Comment by MikeInnes** [comment]
> I fear that we're going to end up having a full HTML parser in Base :|

In CommonMark, the intention is that it should be possible to recognise HTML using simple rules, i.e. without a full parser. You don't have to do any processing on it so it's not to tricky to match up `<` and `>` characters and avoid escaping that section.

**Comment by samoconnor** [comment]
I currently use this NodeJS hack as a work around for generating AWS documentation: https://github.com/samoconnor/AWSCore.jl/blob/master/src/HTML2MD.jl

It would be nice if HTML in markdown just worked.

**Comment by clarkevans** [comment]
I don't know if it helps, but there is a lexer I ported from mbostok's htl project, it's at -- https://github.com/MechanicalRabbit/HypertextLiteral.jl/blob/master/src/lexer.jl

Anyway, I don't like this idea, even if Markdown supports it.  I'd prefer we use Julia interpolation. https://github.com/JuliaDocs/Documenter.jl/issues/517#issuecomment-866130024

**Comment by fingolfin** [comment]
Basic support for "HTML blocks" (as per section 4.6 of the CommonMark spec) is now in, thanks to PR #60632. But support for "Raw HTML" (as per section 6.6 of the CommonMark spec) is not. The latter is for "inline" use, roughly speaking.

Supporting HTML blocks in CommonMark Markdown actually *can* be done using regex, because they deliberately designed it to do that. The CommonMark is intentionally only supporting a subset of all valid HTML so that it can be "parsed" sufficiently for it to be passed through for a Markdown-to-HTML renderer.

Of course that means there is some HTML that can't be handled. And it also means that rendering to the the REPL / a terminal will *not* render that HTML. But it will (well: could...) appear fine in e.g. a manual generated by Documenter (I've not actually tried that, and I think it'll require some changes to Documenter and/or MarkdownAST -- but I am hopeful it will eventually work).

Anyway, let's put it this way: we can support maybe 90% (scientifically accurate number ;-) ) of all uses with this approach, which is better than supporting 0% as in the current setup. And it's CommonMark compliant. Well, the general approach is -- there are all kinds of corner cases were we are no compliant, and may never be; we are still only passing 297 of 652 test of the CommonMark spec test suite. But considering we were at 128 passing tests when I started on this a week ago, that's not *too* bad.

Anyway: HTML entities are coming in PR #60629. Basic "Raw HTML" support will also come.

For now I am really using the regex approach but for efficiency and to handle corner cases I might switch to a simple recursive descent parser for the "grammar" of open/closing tags and attributes.
